apps/aevm/src/aevm_eeevm.erl

```diff
@@ -1424,3 +1408,92 @@ log_topics(AccountAddress, Topics) ->
 create_account(_Value, _CodeArea, State) ->
     %% TODO: Do actual account creation
     {16#DEADC0DE, State}.
+
+
+%% ------------------------------------------------------------------------
+%% CALL
+%% ------------------------------------------------------------------------
+
+recursive_call(State, Op) ->
+    CallDepth = aevm_eeevm_state:calldepth(State),
+    case CallDepth < 1024 of
+        false -> {0, State}; %% TODO: Should this consume gas?
+        true  -> recursive_call(CallDepth, State, Op)
+    end.
+
+recursive_call(CallDepth, StateIn, Op) ->
+    %% Message-call into an account.
+    %% i ≡ µm[µs[3] . . .(µs[3] + µs[4] − 1)]
+    State0            = spend_call_gas(StateIn, Op),
+
+    {Gas, State1}     = pop(State0),
+    {To, State2}      = pop(State1),
+    {Value, State3}   = case Op of
+                            ?CALL         -> pop(State2);
+                            ?CALLCODE     -> pop(State2);
+                            ?DELEGATECALL -> {aevm_eeevm_state:value(State2), State2}
+                        end,
+    {IOffset, State4} = pop(State3),
+    {ISize, State5}   = pop(State4),
+    {OOffset, State6} = pop(State5),
+    {OSize, State7}   = pop(State6),
+    Dest              = case Op of
+                            ?CALL -> To;
+                            ?CALLCODE -> aevm_eeevm_state:address(State6);
+                            ?DELEGATECALL -> aevm_eeevm_state:address(State6)
+                        end,
+    {I, State8}       = aevm_eeevm_memory:get_area(IOffset, ISize, State7),
+    GasAfterSpend     = aevm_eeevm_state:gas(State8),
+    Code              = aevm_eeevm_state:extcode(To, State8),
+    CallGas = case Value =/= 0 of
+                  true -> ?GCALLSTIPEND + Gas;
+                  false -> Gas
+              end,
+    case GasAfterSpend >= CallGas of
+        true  -> ok;
+        false -> eval_error(out_of_gas, State7)
+    end,
+    Caller = case Op of
+                 ?CALL -> aevm_eeevm_state:address(State8);
+                 ?CALLCODE -> aevm_eeevm_state:address(State8);
+                 ?DELEGATECALL -> aevm_eeevm_state:caller(State8)
+             end,
+    CallState = aevm_eeevm_state:prepare_for_call(Caller, Dest, CallGas, Value,
+                                                  Code, CallDepth,
+                                                  State8),
+    case aevm_eeevm_state:no_recursion(State8) of
+        true  -> %% TODO: Set callcreates
+            GasOut = GasAfterSpend + CallGas,
+            State9 = aevm_eeevm_state:set_gas(GasOut, State8),
+            State10 = aevm_eeevm_state:add_callcreates(#{ data => I
+                                                        , destination => Dest
+                                                        , gasLimit => CallGas
+                                                        , value => Value
+                                                        }, State9),
+            {1, State10};
```

Comment:
I don't quite understand this case. What is the purpose of the `no_recursion` flag? If we're not performing the call (due to `no_recursion`) the return memory will contain garbage. At the very least I would expect the return value to be `0`.

Reply:
As I remember it, the no_recursion flag is for testing. The add_callcreates is adding some test data that is verified in the test framework.

Reply:
Is this something we need? It would make more sense to use a different implementation of the chain API if we don't want to perform any calls but still make sure that they happen. That way we can also test contracts that actually look at the return value of a call.

Reply:
It is just the way the imported test suite works. I guess they are aiming at testing the setup of calls.

I agree that there probably are better ways of testing it.

Reply:
The documentation for the vmTests states that:
  "Since these tests are meant only as a basic test of VM operation, the ``CALL`` and 
``CREATE`` instructions are not actually executed. To provide the possibility of 
testing to guarantee they were actually run at all, a separate portion ``callcreates`` 
details each ``CALL`` or ``CREATE`` operation in the order they would have been executed. 
Furthermore, gas required is simply that of the VM execution: the gas cost for 
transaction processing is excluded."

But that turns out to not be completely true. Some of the vmTests actually performs the call.
Possibly one can determine this on whether the callcreats is [] or not in the spec.
I fixed the test suite so that it sends no_recursion = false for the tests that I have discovered that does the calls.